graylog2-web-interface/src/components/search/MessageDetail.tsx

```diff
@@ -0,0 +1,194 @@
+  messageStreams: Array<Stream>,
+  streamIds: Immutable.Set<string>,
+  streams: Immutable.Map<string, Stream>
+}) => {
+  if (messageStreams) {
+    return (
+      <>
+        {messageStreams.map((stream) => (<li key={stream.id}><StreamLink stream={stream} /></li>))}
+      </>
+    );
+  }
+
+  return (
+    <>
+      {streamIds.map((id) => streams.get(id))
+        .filter((stream) => !!stream)
+        .map((stream) => <li key={stream.id}><StreamLink stream={stream} /></li>)}
+    </>
+  );
+};
+
+type Props = {
+  message: Message & { streams?: Array<Stream> },
+  inputs?: Immutable.Map<string, Input>,
+  nodes?: Immutable.Map<string, { short_node_id: string, hostname: string }>,
+  streams?: Immutable.Map<string, Stream>,
+  renderForDisplay: (fieldName: string) => React.ReactNode,
+  customFieldActions?: React.ReactNode
+}
+
+const MessageDetail = ({ renderForDisplay, inputs, nodes, streams, message, customFieldActions }: Props) => {
+  const streamIds = Immutable.Set(message.stream_ids);
+  const rawTimestamp = message.fields.timestamp;
+  const timestamp = [
+    <dt key={`dt-${rawTimestamp}`}>Timestamp</dt>,
+    <dd key={`dd-${rawTimestamp}`}><Timestamp dateTime={rawTimestamp} /></dd>,
+  ];
+
+  return (
+    <div>
+      <Row className="row-sm">
+        <Col md={12}>
+          {customFieldActions
+            ? <MessageActions messageIndex={message.index} messageId={message.id} />
+            : <ButtonGroup className="pull-right" />}
+          <MessageDetailsTitle>
+            <Icon name="mail" />
+            &nbsp;
+            {message.index ? (
+              <Link to={Routes.message_show(message.index, message.id)}>
+                {message.id}
+              </Link>
+            ) : <span>{message.id} <Label bsStyle="warning">Not stored</Label></span>}
+          </MessageDetailsTitle>
+        </Col>
+      </Row>
+      <Row>
+        <Col md={3}>
+          <MessageDetailsDefinitionList>
+            {timestamp}
+            {(message.source_input_id && message.source_node_id && nodes) && (
+              <div>
+                <dt>Received by</dt>
+                <dd>
+                  <em><InputName inputs={inputs} inputId={message.source_input_id} /></em>{' '}
+                  on <NodeName nodes={nodes} nodeId={message.source_node_id} />
+
+                  {/* Legacy */}
+                  {message.source_radio_id && (
+                    <>
+                      <br />
+                      <span>
+                        via <em><InputName inputs={inputs} inputId={message.source_radio_input_id} /></em> on
+                        radio <NodeName nodes={nodes} nodeId={message.source_radio_id} />
+                      </span>
+                    </>
+                  )}
+                </dd>
+              </div>
+            )}
+
+            <dt>Stored in index</dt>
+            <dd>{message.index ? message.index : 'Message is not stored'}</dd>
+
+            {!!streamIds.size && (
```

Comment:
I find either:
```suggestion
            {streamIds.size > 0 && (
```

or

```suggestion
            {!streamIds.isEmpty() && (
```

to be more expressive than the double-bang.